### PR TITLE
Extend and fix RecHitTools

### PR DESCRIPTION
**Pull request by clelange**
- add bhOffset_,
- fix getLayerWithOffset for new geometry,
- add firstLayerBH(), lastLayerBH()

#### PR description:

This PR introduces some important functionality and fixes in/to the RecHitTools required to eventually remove magic numbers in other places of the HGCal reconstruction code.
Related to https://github.com/cms-sw/cmssw/issues/26225

#### PR validation:

Ran some test code with D41 geometry and got the following output:

```text
bhOffset_: 36
EE layers: 28
FH layers: 22
BH layers: 14
first EE layer: 1
first FH layer: 1
first BH layer: 9
EE layer at z = 10000: 28
FH layer at z = 10000: 24
BH layer at z = 10000: 24
lastLayerEE: 28
lastLayerFH: 50
lastLayerBH: 50
firstLayerBH: 37
```

There are some discrepancies that need to be understood, which, however, seem to be in the geometry, underlining the importance of this PR for geometry validation.
For example: D41 should have 50 HGCal layers only, 22 FH layers, 14 BH layers.
After talking to @bsunanda , the error might also be in my setup and there are no issues in D41 geoemtry. We will investigate, but that's independent of the PR.

### CODE REVIEW

**Comment by kpedro88** [comment]
please test

**Comment by clelange** [comment]
OK, that's somewhat expected given the layer numbers I gave above. I'll run those workflows myself to figure out why exactly they fail.

**Comment by clelange** [comment]
After `git cms-checkdeps -a` and adjusting all related packages to use _RecHitTools_ (see the two commits https://github.com/cms-sw/cmssw/pull/27128/commits/bc39efc202c85c70a6698713b62f7fd3a0b8b20e and https://github.com/cms-sw/cmssw/pull/27128/commits/9798a95a94c67e065b43da62aafb38bc5a36489d), I now get a crash due to excessive memory usage, but I'm not sure why this is happening... Will pick this up again next week, but it would help a lot of someone had a look to see if I made an obvious mistake.

**Comment by kpedro88** [comment]
hold
(while memory issue is being resolved - I'm going to try to check it today)

**Comment by clelange** [comment]
The memory issues are fixed now, thanks @kpedro88  for your help!
I locally tested the two workflows that previously failed, and they are fine now.
This PR should be ready for tests and review.

**Comment by kpedro88** [comment]
unhold
@clelange can you apply the suggested code checks?

File: RecoLocalCalo/HGCalRecProducers/interface/HGCalClusteringAlgoBase.h [hunk 1]
```diff
@@ -65,20 +65,25 @@ enum VerbosityLevel { pDEBUG = 0, pWARNING = 1, pINFO = 2, pERROR = 3 };
...
+   getEventSetupPerAlgorithm(es);
  }
  inline void setVerbosity(VerbosityLevel the_verbosity) {
    verbosity_ = the_verbosity;
  }
  inline void setAlgoId(reco::CaloCluster::AlgoId algo) {algoId_ = algo;}
 
  //max number of layers
- static const unsigned int maxlayer = 52;
+ unsigned int maxlayer;
```
Comment (kpedro88) [inline RecoLocalCalo/HGCalRecProducers/interface/HGCalClusteringAlgoBase.h]: these member variables should have a trailing underscore

**Comment by kpedro88** [comment]
please test

**Comment by kpedro88** [comment]
@clelange are the comparison changes expected/intended?

**Comment by clelange** [comment]
@kpedro88 Since I removed a few hardcoded values, which are different e.g. for the D41 geometry, some changes are expected. See e.g. RecoLocalCalo/HGCalRecProducers/interface/HGCalClusteringAlgoBase.h where `lastLayerFH` is now not 40 anymore but 50. Additionally, the thresholds in RecoLocalCalo/HGCalRecProducers/plugins/HGCalImagingAlgo.cc and other places still assume that EE-FH-BH are separated by layer, which is also not true anymore for the new geometry, but only EE is separate from the other two. Therefore, I expect comparison changes, but only for events with activity in the hadronic parts of the calorimeter.

**Comment by kpedro88** [comment]
Well, this gets into the distinction raised in #26225 (which is not currently followed in many places). lastLayerFH should still be 40 if it refers to the fine sampling hadronic section, but it should be 50 if it refers to the silicon part of the hadronic section.

Also, there are now some merge conflicts.

**Comment by clelange** [comment]
If I understand the new geometry correctly, this distinction is not by layer, but it's by wafer/sensor. There are fine and coarse sections in each layer of the `HGCalSi` part (what we refer to as _FH_). This means that the code logic needs to be changed such that one does not check the layer, but the wafer/sensor size instead. According to @bsunanda this is currently the only way of checking this---he might add additional functionality at a later point.
Since `HGCalImagingAlgo` and `HGCal3DClustering` will be replaced by `HGCalCLUE` and further TICL developments, I'm not sure it is worth the effort to adjust the code for the former two, and for the latter optimisation studies need to be performed to find new parameters/thresholds, which would require further code changes anyway. What do you think?

I'll rebase onto tonights IB tomorrow to solve the merge conflicts. I'll then also remove the (afaik last remaining) hardcoded values in https://github.com/cms-sw/cmssw/blob/921fbc5bf3a5d3e43f1a89bb621f5b2f03a388ed/RecoLocalCalo/HGCalRecAlgos/interface/HGCalRecHitSimpleAlgo.h#L50-L67

**Comment by kpedro88** [comment]
I think this is part of the confusion. The "front part" of the hadronic section (FH) is mostly silicon, but all fine-grained. The "back part" (BH) is part silicon and part scintillator, but all coarse-grained. The fine/coarse boundary is still fixed at a specific z value; it's just the material boundaries that are now based on radiation contours. We shouldn't refer to the HGCalSi as FH, because it covers parts of both FH and BH. But if the usage here is only meant for HGCalSi, then the numerical value is at least correct.

I agree we shouldn't spend too much effort on outdated algorithms.

Getting rid of the hardcoded values in HGCalRecHitSimpleAlgo.h would be great (but will conflict with #27129...).

**Comment by bsunanda** [comment]
Front 8 layers of HE part are all silicon and then one has mixed layers of silicon and scintillators. However the coarse and fine part in a given layer depend on R-value and we have one function in HGCalDDDConstant which gives the type (0, 1, or 2) with inputs u, v and layer number. The function will be available in one of the PR's which is not yet integrated

**Comment by clelange** [comment]
OK, so as a simple mitigation I can just implement the BH offset as replacement for `lastLayerFH()` so that the first 8 layers of the hadronic endcap get the _FH_ thresholds and the ones behind the _BH_ (previously all scintillator) thresholds as a starting point for someone to do the optimisation.

Then I'll leave `HGCalRecHitSimpleAlgo.h` untouched for now and we can clean it up once https://github.com/cms-sw/cmssw/pull/27129 is merged.

**Comment by perrotta** [comment]
hold
(Besides needing a rebase even now, this PR would further conflict with #27026 in ```RecoLocalCalo/HGCalRecAlgos/src/RecHitTools.cc```. Unless this one is urgent, we will wait for #27026 to go in first.)

**Comment by clelange** [comment]
@kpedro88 @bsunanda I've resolved the merge conflicts and updated the clustering code to use the first layer of BH as threshold for the clustering.

@perrotta Most changes in #27160 that conflict with this PR are just code formatting, so this is in my opinion very annoying to rebase and fix all the merge conflicts. I would suggest to remove the code formatting changes from #27160 and then run the code-formatting at a later stage or merge this PR first.

**Comment by perrotta** [comment]
> @perrotta Most changes in #27160 that conflict with this PR are just code formatting, so this is in my opinion very annoying to rebase and fix all the merge conflicts. I would suggest to remove the code formatting changes from #27160 and then run the code-formatting at a later stage or merge this PR first.

Thank you @clelange for the updates and the rebase.
Clang formatting has to be applied at some point, and it will always interfere with some other ongoing development. You guys are very active in HGCal in these days, and some of these PRs will inevitably conflict with some other one: that's by the way one of the reasons for which we have called for a review of the ongoing works in HGCal reco to be discussed in the next reco meeting.
As of now, we would like to go as quickly as possible towards merging #27160 (which had to be rebased a few times, already), unless there are urgencies for any of the other competitor PRs, in which case we will try to setup a different schedule.

**Comment by perrotta** [comment]
@cmsbuild please test

**Comment by clelange** [comment]
I've run the code formatting in order to make eventual merging with #27160 easier (hopefully).

**Comment by perrotta** [comment]
please test

**Comment by perrotta** [comment]
unhold

**Comment by kpedro88** [comment]
+upgrade

**Comment by perrotta** [comment]
@clelange 

Having updated the clang format here makes quite hard to review the code and disentangle the "real" updates made in this PR.

Further, this PR will conflict with #27049, which is basically ready to go (and once merged will simplify also the modifications needed by this PR, as there will be a few files that are now modified here and which will simply get removed).

My suggestion is:
- Wait for #27049 being merged, as well as #27129
- Rebase this one, and if #27129 has been already merged you can also update HGCalRecHitSimpleAlgo.h here
- Then there will be a race between this PR and  #27160, and the second one being merged will likely have to rebase once again (maybe not, since you already applied clang-format here... this will have to be verified).

**Comment by clelange** [comment]
@perrotta 

Fine with me to wait for #27049 and #27129 - I'm watching both of them. If it helps simplify the review, I can also just undo the clang format by moving one commit back, you review the changes, and then I apply clang format again before the merge (and after the tests). This would eventually help #27160 getting merged. Let me know.

**Comment by perrotta** [comment]
Clemens Lange <notifications@github.com> ha scritto:

> @perrotta
>
> Fine with me to wait for #27049 and #27129 - I'm watching both of  
> them. If it helps simplify the review, I can also just undo the  
> clang format by moving one commit back, you review the changes, and  
> then I apply clang format again before the merge (and after the  
> tests). This would eventually help #27160 getting merged. Let me know.

Indeed that would help...
Thank you,
Andrea

**Comment by clelange** [comment]
OK, no problem, I've reverted the last commit, and can then re-apply clang format after the code review.

**Comment by perrotta** [comment]
@clelange : I'm fine with the code review now, and also with the comparison results, as all differences were already addressed by you and @kpedro88 earlier on in this thread. Please go ahead with the clang formatting, which I think it is nowadays necessary to restart the tests.

Well, the natural order of things would be to merge #27129 first and then also update HGCalRecHitSimpleAlgo.h here. @fabiocos please let us know which are the plans for #27129, which is already fully signed by now: please notice that there is also #27160 in the queue, waiting for this PR being merged.

**Comment by clelange** [comment]
OK, I'll get to it, just rebasing onto the latest IB and then looking into `HGCalRecHitSimpleAlgo.h`. I'll let you know once done.

**Comment by fabiocos** [comment]
@perrotta @clelange #27129 has been integrated, so now you may update this PR accordingly

**Comment by clelange** [comment]
@perrotta @kpedro88 I've rebased my changes on top of `official-cmssw/master` to merge with the nose-detector changes by @bsunanda. The `HGCalRecHitSimpleAlgo.h` now also uses the RecHitTools inheriting from https://github.com/cms-sw/cmssw/blob/8f2c1e4dfe8695aca4d3d238114b00e4d6405917/RecoLocalCalo/HGCalRecProducers/interface/HGCalClusteringAlgoBase.h#L86 (I've added it there).
The only non-formatting changes in `HGCalClusteringAlgoBase.h`, `HGCalRecHitSimpleAlgo.h` and `HGCalRecHitWorkerSimple.cc` are related to the use of RecHitTools, see the last two commits.

**Comment by perrotta** [comment]
please test

**Comment by perrotta** [comment]
The tests show now a much larger disagreement with the baseline than the previous version of this PR, in some case even quite significant. See for example in wf 27434.0:

image

which was

image

What is changed is the baseline: therefore, I suspect that you are reverting some change that was applied in the already merged pull requests.

@clelange , could you please check?

**Comment by kpedro88** [comment]
@perrotta I rather suspect that rebasing to master means that the changes from another PR were included on top of this one, let's test again

**Comment by kpedro88** [comment]
please test

**Comment by kpedro88** [comment]
@clelange okay, in this case I would try to make sure the exact previous behavior is preserved

File: RecoLocalCalo/HGCalRecAlgos/interface/HGCalRecHitSimpleAlgo.h [hunk 2]
```diff
@@ -19,70 +19,48 @@
...
-    } else {
-      throw cms::Exception("InvalidRecHit")
-	<< "HGCalRecHitSimpleAlgo encountered a non-HGCal det id: " << baseid.det() << ' ' << baseid.subdetId() << ' ' << baseid.rawId();
-    }
 
-    //    float clockToNsConstant = 25;    
-    float energy = (hfnose ? (uncalibRH.amplitude() * weightsNose_[layer] * 0.001f) :
-                    (uncalibRH.amplitude() * weights_[layer] * 0.001f));
-    float time   = uncalibRH.jitter();
+    if (DetId::Forward == baseid.det() && HFNose == baseid.subdetId())
```
Comment (perrotta) [inline RecoLocalCalo/HGCalRecAlgos/interface/HGCalRecHitSimpleAlgo.h]:
```C++
    bool hfnose = DetId::Forward == baseid.det() && HFNose == baseid.subdetId();
```

**Review by perrotta — COMMENTED** [review]
Only in case you are going to modify anything else ....

File: RecoLocalCalo/HGCalRecAlgos/interface/HGCalRecHitSimpleAlgo.h (hunk 2 above)
Comment (clelange) [reply RecoLocalCalo/HGCalRecAlgos/interface/HGCalRecHitSimpleAlgo.h]: Done

**Comment by clelange** [comment]
@perrotta @kpedro88 With some help from @rovere I fixed the use of RecHitTools in the RecHit calibration algorithms (they were not initialised due to the somewhat weird inheritence structure) so that they output is now equivalent to before. I tested this for one geometry, the central tests should hopefully confirm this for the others as well.

**Comment by kpedro88** [comment]
please test

**Comment by perrotta** [comment]
+1
- It allows to centrally retrieve HGCal layer bounds, and fixes offsets for them
- The changes in the output of jenkis tests show the effect of those fixes in the affected Phase2 geometries, as confirmed in the github thread starting from https://github.com/cms-sw/cmssw/pull/27128#issuecomment-500437990
- Further differences that showed us during the developing phase of this PR were found to be spurious and fixed with the last commit https://github.com/cms-sw/cmssw/pull/27128/commits/68b182e955207f2114f5f6ef8c97516007dfce8a
   - Jenkins tests do not show unexpected differences wrt the baseline any more
- This PR will likely conflict with #27160, which is also supposed to get ready soon

**Comment by perrotta** [comment]
@kpedro88 please have a look

**Comment by kpedro88** [comment]
+upgrade

**Comment by rovere** [comment]
Now that all the mandatory signatures are there, it would be nice integrating this asap, so that #27160 could be rebased for the last time, hopefully, and quickly reviewed and integrated itself.
Thanks.

**Comment by fabiocos** [comment]
@rovere indeed there will be a conflict between this PR and #27160, independent on the relative order of integration. At this point I will integrate it in next IB

**Comment by rovere** [comment]
@fabiocos Yes, I was suggesting to put this in soon, so that the bigger #27160 could be fixed and reviewed. Thanks for the support in this 'messy' period.

**Comment by fabiocos** [comment]
+1

**Pull request merged**